src/main/java/org/apache/maven/plugins/clean/Cleaner.java

```diff
@@ -115,9 +135,102 @@ public void delete( File basedir, Selector selector, boolean followSymlinks, boo
+        if ( fastDir.getAbsolutePath().startsWith( baseDir.getAbsolutePath() ) )
+        {
+            try
+            {
+                File tmpDir = createTempDir( baseDir.getParentFile(), baseDir.getName() + "." );
+                try
+                {
+                    Files.move( baseDir.toPath(), tmpDir.toPath(), StandardCopyOption.ATOMIC_MOVE );
+                    if ( session != null )
+                    {
+                        session.getRequest().getData().put( LAST_DIRECTORY_TO_DELETE, baseDir );
+                    }
+                    baseDir = tmpDir;
+                }
+                catch ( IOException e )
+                {
+                    tmpDir.delete();
+                    if ( logDebug != null )
+                    {
+                        logDebug.log( "Unable to fast delete directory: " + e );
+                    }
+                    return false;
+                }
+            }
+            catch ( IOException e )
+            {
+                if ( logDebug != null )
+                {
+                    logDebug.log( "Unable to fast delete directory: " + e );
+                }
+                return false;
+            }
+        }
+        // Create fastDir and the needed parents if needed
+        if ( fastDir.mkdirs() || fastDir.isDirectory() )
+        {
+            try
+            {
+                File tmpDir = createTempDir( fastDir, "" );
+                File dstDir = new File( tmpDir, baseDir.getName() );
+                // Note that by specifying the ATOMIC_MOVE, we expect an exception to be thrown
+                // if the path leads to a directory on another mountpoint.  If this is the case
+                // or any other exception occurs, an exception will be thrown in which case
+                // the method will return false and the usual deletion will be performed.
+                Files.move( baseDir.toPath(), dstDir.toPath(), StandardCopyOption.ATOMIC_MOVE );
+                BackgroundCleaner.delete( this, tmpDir );
+                return true;
+            }
+            catch ( IOException e )
+            {
+                if ( logDebug != null )
+                {
+                    logDebug.log( "Unable to fast delete directory: " + e );
+                }
+            }
+        }
+        else
+        {
+            if ( logDebug != null )
+            {
+                logDebug.log( "Unable to fast delete directory as the path "
+                        + fastDir + " does not point to a directory or can not be created" );
+            }
+        }
+        return false;
+    }
+
+    private File createTempDir( File baseDir, String prefix ) throws IOException
+    {
+        for ( int i = 0; i < 10; i++ )
+        {
+            int rnd = ThreadLocalRandom.current().nextInt();
+            File tmpDir = new File( baseDir, prefix + Integer.toHexString( rnd ) );
```

Comment:
What's wrong with: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createTempDirectory-java.nio.file.Path-java.lang.String-java.nio.file.attribute.FileAttribute...-?

Reply:
You're right.  Initially I did not want to create the directory, but the code has changed so it definitely makes sense to use the `createTempDirectory` now.

Reply:
I refactored the fastDelete method to use NIO `Path` instead of `File`.